File: python/src/iap/lifecycle/gates.py
"""Lifecycle gates — pure, deterministic ``(alpha_id, evidence) -> GateResult``.

Every gate is one row of :data:`GATE_SPECS`: a name, the evidence block it
reads, the metric it extracts, the comparison and the config key holding its
threshold.  A port implements the table, not eighteen classes.

Comparison kinds (pinned):

* ``min``  — ``value >= threshold`` (inclusive);
* ``max``  — ``value <= threshold`` (inclusive);
* ``gt``   — ``value >  threshold`` (strict; the net-P&L gate mirrors
  ``iap.validation.validate``'s ``net_pnl_1x > 0.0``);
* ``bool`` — the metric itself; ``value`` and ``threshold`` are ``null``.

A gate whose evidence block is absent, or whose metric is ``None``, FAILS
with ``value = null`` (a missing number never passes); the machine decides
separately whether an absent block counts as a failure or as silence
(``iap.lifecycle.machine``).

Gate inventory (config key in ``configs/strategies/lifecycle.json``):

==========================  ==========  =========================================  ==========================
gate                        block       metric                                     threshold key
==========================  ==========  =========================================  ==========================
ledger_entry_exists         research    n_experiments_in_ledger (min)              min_experiments_in_ledger
leakage_clean               research    leakage_passed (bool)                      —
oos_ic                      research    ic (min)                                   min_oos_ic
statistical_significance    research    t_stat (min)                               min_nw_tstat
fold_consistency            research    fold_consistency (min)                     min_fold_sign_consistency
fold_count                  research    n_folds (min)                              min_folds
hypothesis_sign             research    hypothesis_sign_confirmed is True (bool)   —
net_pnl_after_costs         research    net_return_bps (gt)                        min_net_return_bps
capacity                    capacity    capacity_usd (min)                         min_capacity_usd
stability                   research    |ic - rank_ic| / max(|ic|, eps) (max)      max_ic_rank_gap
holdout_ic_tracks_research  validation  |holdout_ic - research_ic| (max)           max_holdout_ic_gap
replay_reproducible         validation  replay_hash_match (bool)                   —
cross_language_parity       validation  parity (bool)                              —
paper_min_sessions          paper       n_sessions (min)                           min_paper_sessions
paper_ic_tracking           paper       |realized_ic - research_ic| (max)          max_paper_ic_gap
paper_net_pnl               paper       net_pnl (min)                              min_paper_net_pnl
no_kill_events              paper       n_kill_events (max)                        max_kill_events
rolling_ic                  live        rolling_ic (min)                           watch_ic_gate (strategies)
==========================  ==========  =========================================  ==========================

**Why the Pearson/rank gap is the stability rule.**  Everything a research
result carries is a point estimate over one out-of-sample window; the one
pair of numbers that measures the *shape* of the signal-label relation
rather than its strength is (``ic``, ``rank_ic``): Pearson weights every row
by its magnitude, Spearman weights every row equally.  A rank IC far above
the Pearson IC means the monotone relation exists but the linear score is
diluted by outliers; a rank IC near zero under a healthy Pearson IC means
the IC comes from a handful of extreme rows.  Both are instabilities a live
book will feel first.  ``|ic - rank_ic| / max(|ic|, eps) <= 1.0`` therefore
requires the rank IC to lie in ``[0, 2 * ic]`` for a positive IC (same sign,
at most twice the linear IC) — computable from any ``ExperimentResult``,
scale-free, and with a threshold a port can state in one line.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Callable, Dict, Mapping, Optional, Tuple, Union

from iap.contracts.types import GateResult
from iap.lifecycle.config import PolicyConfig
from iap.lifecycle.evidence import Evidence

__all__ = [
    "GATE_SPECS",
    "Gate",
    "GateSpec",
    "build_gates",
    "ic_rank_gap",
]

Metric = Union[float, int, bool, None]


def ic_rank_gap(ic: float, rank_ic: float, eps: float) -> float:
    """``|ic - rank_ic| / max(|ic|, eps)`` — the pinned stability statistic."""
    return abs(ic - rank_ic) / max(abs(ic), eps)


def _research(field: str) -> Callable[[Evidence, PolicyConfig], Metric]:
    def read(ev: Evidence, cfg: PolicyConfig) -> Metric:
        return None if ev.research is None else getattr(ev.research, field)
    return read


def _hypothesis(ev: Evidence, cfg: PolicyConfig) -> Metric:
    if ev.research is None:
        return None
    return ev.research.hypothesis_sign_confirmed is True


def _capacity(ev: Evidence, cfg: PolicyConfig) -> Metric:
    return ev.capacity_usd


def _stability(ev: Evidence, cfg: PolicyConfig) -> Metric:
    if ev.research is None:
        return None
    return ic_rank_gap(ev.research.ic, ev.research.rank_ic, cfg.gates.ic_rank_gap_eps)


def _validation(field: str) -> Callable[[Evidence, PolicyConfig], Metric]:
    def read(ev: Evidence, cfg: PolicyConfig) -> Metric:
        return None if ev.validation is None else getattr(ev.validation, field)
    return read


def _holdout_gap(ev: Evidence, cfg: PolicyConfig) -> Metric:
    if ev.validation is None:
        return None
    return abs(ev.validation.holdout_ic - ev.validation.research_ic)


def _paper(field: str) -> Callable[[Evidence, PolicyConfig], Metric]:
    def read(ev: Evidence, cfg: PolicyConfig) -> Metric:
        return None if ev.paper is None else getattr(ev.paper, field)
    return read


def _paper_gap(ev: Evidence, cfg: PolicyConfig) -> Metric:
    if ev.paper is None:
        return None
    return abs(ev.paper.realized_ic - ev.paper.research_ic)


def _rolling_ic(ev: Evidence, cfg: PolicyConfig) -> Metric:
    if ev.live is None or not ev.live.informative:
        return None
    return ev.live.rolling_ic


def _threshold_from_gates(key: str) -> Callable[[PolicyConfig], Optional[float]]:
    def read(cfg: PolicyConfig) -> Optional[float]:
        return float(getattr(cfg.gates, key))
    return read


def _threshold_none(cfg: PolicyConfig) -> Optional[float]:
    return None


def _threshold_watch_gate(cfg: PolicyConfig) -> Optional[float]:
    return cfg.live.watch_ic_gate


@dataclass(frozen=True)
class GateSpec:
    """One row of the gate table (see module docstring)."""

    name: str
    block: str
    kind: str
    threshold_key: Optional[str]
    metric: Callable[[Evidence, PolicyConfig], Metric]
    threshold: Callable[[PolicyConfig], Optional[float]]

    def __post_init__(self) -> None:
        if self.kind not in ("min", "max", "gt", "bool"):
            raise ValueError(f"gate {self.name}: unknown comparison {self.kind!r}")
        if (self.kind == "bool") != (self.threshold_key is None):
            raise ValueError(f"gate {self.name}: bool gates have no threshold key")


def _spec(name: str, block: str, kind: str, key: Optional[str],
          metric: Callable[[Evidence, PolicyConfig], Metric]) -> GateSpec:
    if key is None:
        threshold = _threshold_none
    elif key == "watch_ic_gate":
        threshold = _threshold_watch_gate
    else:
        threshold = _threshold_from_gates(key)
    return GateSpec(name=name, block=block, kind=kind, threshold_key=key,
                    metric=metric, threshold=threshold)


#: The complete gate table, in a fixed order (the machine evaluates the
#: subset of an edge in the edge's own pinned order).
GATE_SPECS: Tuple[GateSpec, ...] = (
    _spec("ledger_entry_exists", "research", "min", "min_experiments_in_ledger",
          _research("n_experiments_in_ledger")),
    _spec("leakage_clean", "research", "bool", None, _research("leakage_passed")),
    _spec("oos_ic", "research", "min", "min_oos_ic", _research("ic")),
    _spec("statistical_significance", "research", "min", "min_nw_tstat",
          _research("t_stat")),
    _spec("fold_consistency", "research", "min", "min_fold_sign_consistency",
          _research("fold_consistency")),
    _spec("fold_count", "research", "min", "min_folds", _research("n_folds")),
    _spec("hypothesis_sign", "research", "bool", None, _hypothesis),
    _spec("net_pnl_after_costs", "research", "gt", "min_net_return_bps",
          _research("net_return_bps")),
    _spec("capacity", "capacity", "min", "min_capacity_usd", _capacity),
    _spec("stability", "research", "max", "max_ic_rank_gap", _stability),
    _spec("holdout_ic_tracks_research", "validation", "max", "max_holdout_ic_gap",
          _holdout_gap),
    _spec("replay_reproducible", "validation", "bool", None,
          _validation("replay_hash_match")),
    _spec("cross_language_parity", "validation", "bool", None, _validation("parity")),
    _spec("paper_min_sessions", "paper", "min", "min_paper_sessions",
          _paper("n_sessions")),
    _spec("paper_ic_tracking", "paper", "max", "max_paper_ic_gap", _paper_gap),
    _spec("paper_net_pnl", "paper", "min", "min_paper_net_pnl", _paper("net_pnl")),
    _spec("no_kill_events", "paper", "max", "max_kill_events",
          _paper("n_kill_events")),
    _spec("rolling_ic", "live", "min", "watch_ic_gate", _rolling_ic),
)

_SPEC_BY_NAME: Mapping[str, GateSpec] = {s.name: s for s in GATE_SPECS}
if len(_SPEC_BY_NAME) != len(GATE_SPECS):
    raise RuntimeError("duplicate gate name in GATE_SPECS")


class Gate:
    """A :class:`GateSpec` bound to a :class:`PolicyConfig`; satisfies
    ``iap.contracts.protocols.LifecycleGate``."""

    __slots__ = ("_spec", "_config")

    def __init__(self, spec: GateSpec, config: PolicyConfig) -> None:
        self._spec = spec
        self._config = config

    @property
    def name(self) -> str:
        return self._spec.name

    @property
    def spec(self) -> GateSpec:
        return self._spec

    @property
    def threshold(self) -> Optional[float]:
        """The bound threshold (``None`` for a boolean gate)."""
        return self._spec.threshold(self._config)

    def evaluate(self, alpha_id: str, evidence: Any) -> GateResult:
        """Pure: the same evidence always yields the same result.
        ``alpha_id`` is accepted for the protocol; no gate is alpha-specific."""
        if not isinstance(evidence, Evidence):
            raise TypeError(f"gate {self.name}: expected Evidence, got "
                            f"{type(evidence).__name__}")
        metric = self._spec.metric(evidence, self._config)
        threshold = self.threshold
        if self._spec.kind == "bool":
            return GateResult(passed=metric is True, value=None, threshold=None)
        if metric is None:
            return GateResult(passed=False, value=None, threshold=threshold)
        value = float(metric)
        if self._spec.kind == "min":
            passed = value >= threshold
        elif self._spec.kind == "max":
            passed = value <= threshold
        else:
            passed = value > threshold
        return GateResult(passed=passed, value=value, threshold=threshold)

    def __repr__(self) -> str:
        return f"Gate({self.name!r}, {self._spec.kind}, threshold={self.threshold!r})"


def build_gates(config: PolicyConfig) -> Dict[str, Gate]:
    """Every gate of the table bound to ``config``, keyed by name (insertion
    order = table order)."""
    return {spec.name: Gate(spec, config) for spec in GATE_SPECS}

File: java/src/main/java/com/iap/lifecycle/Gates.java
package com.iap.lifecycle;

import java.util.function.BiFunction;
import java.util.function.Function;

/**
 * The lifecycle gate table ({@code iap.lifecycle.gates.GATE_SPECS}): every
 * gate is one row — name, evidence block, metric, comparison kind and the
 * config key holding its threshold. Comparison kinds (pinned):
 * {@code MIN} ⇒ {@code value >= threshold}; {@code MAX} ⇒
 * {@code value <= threshold}; {@code GT} ⇒ {@code value > threshold}
 * (strict); {@code BOOL} ⇒ the metric itself with {@code value} and
 * {@code threshold} {@code null}. A gate whose evidence block is absent, or
 * whose metric is absent, FAILS with {@code value = null} (a missing number
 * never passes); the machine decides separately whether an absent block
 * counts as a failure or as silence.
 *
 * <p>The stability rule {@code |ic - rank_ic| / max(|ic|, eps) <= max_ic_rank_gap}
 * requires the rank IC to lie in {@code [0, 2 * ic]} for a positive IC: same
 * sign, at most twice the linear IC — the one pair of numbers in an
 * experiment result that measures the shape of the signal–label relation
 * rather than its strength.
 */
public enum Gates {
    LEDGER_ENTRY_EXISTS("ledger_entry_exists", "research", Kind.MIN,
            "min_experiments_in_ledger",
            (ev, c) -> ev.research() == null ? null
                    : (double) ev.research().nExperimentsInLedger(),
            c -> (double) c.gates().minExperimentsInLedger()),
    LEAKAGE_CLEAN("leakage_clean", "research", Kind.BOOL, null,
            (ev, c) -> ev.research() == null ? null : bool(ev.research().leakagePassed()),
            null),
    OOS_IC("oos_ic", "research", Kind.MIN, "min_oos_ic",
            (ev, c) -> ev.research() == null ? null : ev.research().ic(),
            c -> c.gates().minOosIc()),
    STATISTICAL_SIGNIFICANCE("statistical_significance", "research", Kind.MIN,
            "min_nw_tstat",
            (ev, c) -> ev.research() == null ? null : ev.research().tStat(),
            c -> c.gates().minNwTstat()),
    FOLD_CONSISTENCY("fold_consistency", "research", Kind.MIN,
            "min_fold_sign_consistency",
            (ev, c) -> ev.research() == null ? null : ev.research().foldConsistency(),
            c -> c.gates().minFoldSignConsistency()),
    FOLD_COUNT("fold_count", "research", Kind.MIN, "min_folds",
            (ev, c) -> ev.research() == null ? null : (double) ev.research().nFolds(),
            c -> (double) c.gates().minFolds()),
    HYPOTHESIS_SIGN("hypothesis_sign", "research", Kind.BOOL, null,
            (ev, c) -> ev.research() == null ? null
                    : bool(Boolean.TRUE.equals(ev.research().hypothesisSignConfirmed())),
            null),
    NET_PNL_AFTER_COSTS("net_pnl_after_costs", "research", Kind.GT,
            "min_net_return_bps",
            (ev, c) -> ev.research() == null ? null : ev.research().netReturnBps(),
            c -> c.gates().minNetReturnBps()),
    CAPACITY("capacity", "capacity", Kind.MIN, "min_capacity_usd",
            (ev, c) -> ev.capacityUsd(), c -> c.gates().minCapacityUsd()),
    STABILITY("stability", "research", Kind.MAX, "max_ic_rank_gap",
            (ev, c) -> ev.research() == null ? null
                    : icRankGap(ev.research().ic(), ev.research().rankIc(),
                            c.gates().icRankGapEps()),
            c -> c.gates().maxIcRankGap()),
    HOLDOUT_IC_TRACKS_RESEARCH("holdout_ic_tracks_research", "validation", Kind.MAX,
            "max_holdout_ic_gap",
            (ev, c) -> ev.validation() == null ? null
                    : Math.abs(ev.validation().holdoutIc() - ev.validation().researchIc()),
            c -> c.gates().maxHoldoutIcGap()),
    REPLAY_REPRODUCIBLE("replay_reproducible", "validation", Kind.BOOL, null,
            (ev, c) -> ev.validation() == null ? null
                    : bool(ev.validation().replayHashMatch()),
            null),
    CROSS_LANGUAGE_PARITY("cross_language_parity", "validation", Kind.BOOL, null,
            (ev, c) -> ev.validation() == null ? null : bool(ev.validation().parity()),
            null),
    PAPER_MIN_SESSIONS("paper_min_sessions", "paper", Kind.MIN, "min_paper_sessions",
            (ev, c) -> ev.paper() == null ? null : (double) ev.paper().nSessions(),
            c -> (double) c.gates().minPaperSessions()),
    PAPER_IC_TRACKING("paper_ic_tracking", "paper", Kind.MAX, "max_paper_ic_gap",
            (ev, c) -> ev.paper() == null ? null
                    : Math.abs(ev.paper().realizedIc() - ev.paper().researchIc()),
            c -> c.gates().maxPaperIcGap()),
    PAPER_NET_PNL("paper_net_pnl", "paper", Kind.MIN, "min_paper_net_pnl",
            (ev, c) -> ev.paper() == null ? null : ev.paper().netPnl(),
            c -> c.gates().minPaperNetPnl()),
    NO_KILL_EVENTS("no_kill_events", "paper", Kind.MAX, "max_kill_events",
            (ev, c) -> ev.paper() == null ? null : (double) ev.paper().nKillEvents(),
            c -> (double) c.gates().maxKillEvents()),
    ROLLING_IC("rolling_ic", "live", Kind.MIN, "watch_ic_gate",
            (ev, c) -> ev.live() == null || !ev.live().informative() ? null
                    : ev.live().rollingIc(),
            c -> c.live().watchIcGate());

    /** Comparison kind. */
    public enum Kind {
        MIN, MAX, GT, BOOL
    }

    private static final double BOOL_TRUE = 1.0;

    private final String gateName;
    private final String block;
    private final Kind kind;
    private final String thresholdKey;
    /** Numeric metric (null = absent); BOOL gates encode true as 1.0. */
    private final BiFunction<Evidence, PolicyConfig, Double> metric;
    private final Function<PolicyConfig, Double> threshold;

    Gates(String gateName, String block, Kind kind, String thresholdKey,
            BiFunction<Evidence, PolicyConfig, Double> metric,
            Function<PolicyConfig, Double> threshold) {
        this.gateName = gateName;
        this.block = block;
        this.kind = kind;
        this.thresholdKey = thresholdKey;
        this.metric = metric;
        this.threshold = threshold;
    }

    private static Double bool(boolean b) {
        return b ? BOOL_TRUE : 0.0;
    }

    /** The pinned stability statistic {@code |ic - rank_ic| / max(|ic|, eps)}. */
    public static double icRankGap(double ic, double rankIc, double eps) {
        return Math.abs(ic - rankIc) / Math.max(Math.abs(ic), eps);
    }

    /** Gate name on the wire. */
    public String gateName() {
        return gateName;
    }

    /** Evidence block the gate reads. */
    public String block() {
        return block;
    }

    /** Comparison kind. */
    public Kind kind() {
        return kind;
    }

    /** Config key of the threshold ({@code null} for BOOL gates). */
    public String thresholdKey() {
        return thresholdKey;
    }

    /** The bound threshold ({@code null} for a BOOL gate). */
    public Double threshold(PolicyConfig config) {
        return threshold == null ? null : threshold.apply(config);
    }

    /** Pure: the same evidence always yields the same result. */
    public GateResult evaluate(Evidence evidence, PolicyConfig config) {
        Double m = metric.apply(evidence, config);
        if (kind == Kind.BOOL) {
            return new GateResult(m != null && m == BOOL_TRUE, null, null);
        }
        Double th = threshold(config);
        if (m == null) {
            return new GateResult(false, null, th);
        }
        double value = m;
        boolean passed = switch (kind) {
            case MIN -> value >= th;
            case MAX -> value <= th;
            case GT -> value > th;
            case BOOL -> throw new IllegalStateException("unreachable");
        };
        return new GateResult(passed, value, th);
    }

    /** Look a gate up by wire name. */
    public static Gates byName(String name) {
        for (Gates g : values()) {
            if (g.gateName.equals(name)) {
                return g;
            }
        }
        throw new IllegalArgumentException("unknown lifecycle gate '" + name + "'");
    }
}
